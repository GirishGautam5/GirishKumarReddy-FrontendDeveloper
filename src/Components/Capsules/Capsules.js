import React from "react";
import Pagination from "../Pagination/Pagination";

export const filteredData = (capsules, filter) =>
  capsules?.filter((item) => {
    return (
      item?.capsule_serial?.includes(filter.type) &&
      item?.status?.includes(filter.status) &&
      item?.landings?.toString().includes(filter.landings)
    );
  });

export default function Capsules() {
  const [capsules, setCapsules] = React.useState([]);
  const [currentPage, setCurrentPage] = React.useState(1);
  const [loading, setLoading] = React.useState(true);
  const [filter, setFilter] = React.useState({
    type: "",
    status: "",
    landings: "",
  });

  const handleFilterChange = (event) => {
    const { name, value } = event.target;
    setFilter({ ...filter, [name]: value });
  };

  const capsulesPerPage = 9;
  React.useEffect(() => {
    const fetchCapsulesData = async () => {
      const response = await fetch("https://api.spacexdata.com/v3/capsules");
      const data = await response.json();
      setCapsules(data);
      setLoading(false);
    };
    fetchCapsulesData();
  }, []);

  const filteredCapsulesData = filteredData(capsules, filter);
  const showFilteredData = filter.type || filter.status || filter.landings;
  const displayCapsules = showFilteredData ? filteredCapsulesData : capsules;
  const lastCapsuleIndex = currentPage * capsulesPerPage;
  const firstCapsuleIndex = lastCapsuleIndex - capsulesPerPage;
  const capsulesData = displayCapsules?.slice(
    firstCapsuleIndex,
    lastCapsuleIndex
  );

  return loading ? (
    <div className="loader"></div>
  ) : (
    <>
      <section>
        <h1 className="heading mb-5">Capsules</h1>
        <div className="max-width grid grid-cols-1 gap-5 md:grid-cols-2 lg:grid-cols-3 xl:grid-cols-3 mt-10">
          <label className="text-white">
            Capsule type:{" "}
            <input
              type="text"
              name="type"
              placeholder="Search by Capsule type(C101)"
              value={filter.type}
              onChange={handleFilterChange}
            />
          </label>
          <label className="text-white">
            Capsule Status:{" "}
            <input
              type="text"
              name="status"
              placeholder="Search by status"
              value={filter.status}
              onChange={handleFilterChange}
            />
          </label>
          <label className="text-white">
            Total Landings:{" "}
            <input
              type="text"
              name="landings"
              placeholder="Search by total landings"
              value={filter.landings}
              onChange={handleFilterChange}
            />
          </label>
          {capsulesData.map((capsule) => (
            <article
              key={capsule?.capsule_serial}
              className="articles capsules"
            >
              <h2 className="text-xl font-bold mb-5">
                {capsule?.type},{" "}
                <span className="text-base font-normal opacity-75">
                  {capsule?.capsule_serial}
                </span>
              </h2>
              <ul>
                <li className="mb-1">Reused times: {capsule?.reuse_count}</li>
                <li className="mb-1">landings: {capsule?.landings}</li>
                <li className="mb-1">Details: {capsule?.details || "NA"} </li>
                <li className="mb-1">
                  Original Launch: {capsule?.original_launch || "NA"}
                </li>
                <li className="mb-1">
                  Original Launch Unix: {capsule?.original_launch_unix || "NA"}
                </li>
                <li
                  className={`capitalize ${
                    capsule?.status === "active"
                      ? "text-green-500"
                      : "text-red-500"
                  }`}
                >
                  Status: {capsule?.status}
                </li>
              </ul>
            </article>
          ))}
          {showFilteredData && filteredCapsulesData?.length === 0 && (
            <h1 className="noResult">No Search Results ....</h1>
          )}
        </div>
      </section>
      <Pagination
        totalCapsules={displayCapsules?.length}
        capsulesPerPage={capsulesPerPage}
        currentPage={currentPage}
        setCurrentPage={setCurrentPage}
      />
    </>
  );
}
