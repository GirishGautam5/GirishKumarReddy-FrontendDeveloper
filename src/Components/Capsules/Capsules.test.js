import { shallow } from "enzyme";
import Enzyme from "enzyme";
import Adapter from "enzyme-adapter-react-16";
import Capsules, { filteredData } from "./Capsules";
import React from "react";

Enzyme.configure({ adapter: new Adapter() });

describe("Capsules test suite", () => {
  const capsulesComponent = shallow(<Capsules />);
  const useEffectMock = jest.spyOn(React, "useEffect");
  beforeEach(() => {
    useEffectMock.mockClear();
  });
  it("Capsules component  test case", () => {
    expect(capsulesComponent).toMatchSnapshot();
  });
  it("Capsules test case for useeffect", () => {
    const capsules = [
      {
        capsule_serial: "C102",
        type: "Dragon 1.0",
        reuse_count: 0,
        landings: 0,
        details: "Dragon capsule",
        original_launch: "2010-12-08T15:43:00.000Z",
        original_launch_unix: 1291822980,
        status: "active",
      },
    ];
    const mockUseEffect = () => {
      useEffectMock.mockImplementationOnce((f) => f());
    };
    mockUseEffect();
    mockUseEffect();
    mockUseEffect();
    mockUseEffect();
    React.useState = jest
      .fn()
      .mockReturnValueOnce([capsules, (val) => val])
      .mockReturnValueOnce([1, (val) => val])
      .mockReturnValueOnce([false, (val) => val])
      .mockReturnValueOnce([
        { type: "C102", status: "active", landings: "0" },
        (val) => val,
      ]);
    const capsulesComponent = shallow(<Capsules />);
    expect(capsulesComponent).toBeDefined();
  });
  it("filtereredData test case", () => {
    const capsules = [
      {
        capsule_serial: "C102",
        type: "Dragon 1.0",
        reuse_count: 0,
        landings: 0,
        details: "Dragon capsule",
        original_launch: "2010-12-08T15:43:00.000Z",
        original_launch_unix: 1291822980,
        status: "active",
      },
    ];
    const filter = {
      type: "C102",
      status: "active",
      landings: 0,
    };
    const result = filteredData(capsules, filter);
    expect(result).toEqual(capsules);
  });
  it("filtereredData test case when filter data is empty", () => {
    const capsules = [
      {
        capsule_serial: "C102",
        type: "Dragon 1.0",
        reuse_count: 0,
        landings: 0,
        details: "Dragon capsule",
        original_launch: "2010-12-08T15:43:00.000Z",
        original_launch_unix: 1291822980,
        status: "rejected",
      },
    ];
    const filter = {
      type: "",
      status: "",
      landings: "",
    };
    const mockUseEffect = () => {
      useEffectMock.mockImplementationOnce((f) => f());
    };
    mockUseEffect();
    mockUseEffect();
    mockUseEffect();
    mockUseEffect();
    React.useState = jest
      .fn()
      .mockReturnValueOnce([capsules, (val) => val])
      .mockReturnValueOnce([1, (val) => val])
      .mockReturnValueOnce([false, (val) => val])
      .mockReturnValueOnce([
        { type: "d", status: "b", landings: "" },
        (val) => val,
      ]);
    const capsulesComponent = shallow(<Capsules />);
    expect(capsulesComponent).toBeDefined();
    const result = filteredData(capsules, filter);
    expect(result).toEqual(capsules);
  });
});
