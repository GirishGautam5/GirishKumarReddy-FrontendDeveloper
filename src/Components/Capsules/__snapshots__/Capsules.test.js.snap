// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Capsules test suite Capsules component  test case 1`] = `
<div
  className="loader"
/>
`;

exports[`Capsules test suite Capsules test case for useeffect 1`] = `
<Fragment>
  <section>
    <h1
      className="heading mb-5"
    >
      Capsules
    </h1>
    <div
      className="max-width grid grid-cols-1 gap-5 md:grid-cols-2 lg:grid-cols-3 xl:grid-cols-3 mt-10"
    >
      <label
        className="text-white"
      >
        Capsule type:
         
        <input
          name="type"
          onChange={[Function]}
          placeholder="Search by Capsule type(C101)"
          type="text"
          value="C102"
        />
      </label>
      <label
        className="text-white"
      >
        Capsule Status:
         
        <input
          name="status"
          onChange={[Function]}
          placeholder="Search by status"
          type="text"
          value="active"
        />
      </label>
      <label
        className="text-white"
      >
        Total Landings:
         
        <input
          name="landings"
          onChange={[Function]}
          placeholder="Search by total landings"
          type="text"
          value="0"
        />
      </label>
      <article
        className="articles capsules"
        key="C102"
      >
        <h2
          className="text-xl font-bold mb-5"
        >
          Dragon 1.0
          ,
           
          <span
            className="text-base font-normal opacity-75"
          >
            C102
          </span>
        </h2>
        <ul>
          <li
            className="mb-1"
          >
            Reused times: 
            0
          </li>
          <li
            className="mb-1"
          >
            landings: 
            0
          </li>
          <li
            className="mb-1"
          >
            Details: 
            Dragon capsule
             
          </li>
          <li
            className="mb-1"
          >
            Original Launch: 
            2010-12-08T15:43:00.000Z
          </li>
          <li
            className="mb-1"
          >
            Original Launch Unix: 
            1291822980
          </li>
          <li
            className="capitalize text-green-500"
          >
            Status: 
            active
          </li>
        </ul>
      </article>
    </div>
  </section>
  <Pagination
    capsulesPerPage={9}
    currentPage={1}
    setCurrentPage={[Function]}
    totalCapsules={1}
  />
</Fragment>
`;
